### PR TITLE
FHIR Model and temporarily cached data

### PR DESCRIPTION
**Pull request by fhibf**
## Description
Describe the changes in this PR.

## Related issues
Addresses [issue #].

## Testing
Describe how this change was tested.

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Features/Storage/FhirMemoryCache.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            EnsureArg.IsGt(limitSizeInMegabytes, 0, nameof(name));
+
+            _cache = new MemoryCache(
+                name,
+                new NameValueCollection()
+                {
+                    { "CacheMemoryLimitMegabytes", limitSizeInMegabytes.ToString() },
+                });
+
+            _expirationTime = expirationTime;
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Storage/FhirMemoryCache.cs:43]: Is there an LRU type option? Would LRU or a sliding expiration be better?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs [hunk 2]
```diff
@@ -352,8 +354,13 @@ COMMIT TRANSACTION
                     _resourceTypeToId = resourceTypeToId;
                     _resourceTypeIdToTypeName = resourceTypeIdToTypeName;
                     _searchParamUriToId = searchParamUriToId;
-                    _systemToId = systemToId;
-                    _quantityCodeToId = quantityCodeToId;
+
+                    _systemToId = new FhirMemoryCache<int>("systemToId");
+                    _systemToId.AddRange(systemToId);
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:359]: By loading all these rows first into a local concurrent dictionary, and then copying from that collection into the cache, we will still encounter problems if the system table has too many rows.  Could we add the values into the cache directly in the SQL read loop? And avoid the ConcurrentDictionary object?

File: src/Microsoft.Health.Fhir.Core/Features/Storage/FhirMemoryCache.cs (hunk 1 above)
Comment (fhibf) [reply src/Microsoft.Health.Fhir.Core/Features/Storage/FhirMemoryCache.cs:50]: As discussed, we are implementing the sliding window caching.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs (hunk 2 above)
Comment (fhibf) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:359]: @feordin , you are more than correct!! Let me do that!

File: src/Microsoft.Health.Fhir.Core/Features/Storage/FhirMemoryCache.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+            lock (_lock)
+            {
+                if (_cache.Contains(key))
+                {
+                    return (T)_cache[key];
+                }
+
+                AddInternal(key, value);
+            }
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Storage/FhirMemoryCache.cs:81]: Is there a `_cache.AddOrGetExisting`?

Comment (fhibf) [reply src/Microsoft.Health.Fhir.Core/Features/Storage/FhirMemoryCache.cs]: Yes, there is. And reading the docs, I believe this class can be simpler. PR is updated.

**Review by brendankowitz — APPROVED** [review]

**Pull request merged**